[iOS] Grant media capabilities to the WebContent and GPU processes during capture and playback

aestes opened the pull request:
#### 7b6ea01374235df040c52dfcd986447c7aec9aca
<pre>
[iOS] Grant media capabilities to the WebContent and GPU processes during capture and playback
<a href="https://bugs.webkit.org/show_bug.cgi?id=265588">https://bugs.webkit.org/show_bug.cgi?id=265588</a>
<a href="https://rdar.apple.com/116242683">rdar://116242683</a>

Reviewed by Per Arne Vollan.

This patch introduces MediaCapability and ProcessCapabilityGranter, responsible for granting,
revoking, and managing the activation of a MediaCapability.

A MediaCapability is created each time a WebPageProxy navigates to a new top privately-controlled
domain, and is granted to the page&apos;s GPU and WebContent processes. When media playback or capture
becomes active on the page the capability is activated, and remains active for as long as that page
is capable of playback or capture. Granting and activating these capabilities allows iOS to know the
process hierarchy and domain responsible for playback and capture, and allows the system to, e.g.,
resume the responsible processes when starting playback from Now Playing or account for the
responsible domain in System Status. Media capabilities will serve as a replacement to SPIs like
AVSystemController_PIDToInheritApplicationStateFrom and -[AVAudioSession setHostProcessAttribution:].

This feature is behind an off-by-default web preference (MediaCapabilityGrantsEnabled). Adding API
tests is tracked by <a href="https://rdar.apple.com/115750175">rdar://115750175</a>. It wasn&apos;t done in this patch because tests are blocked by
other work.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Added a web preference.
* Source/WebKit/Platform/Logging.h: Added a logging channel.

* Source/WebKit/Platform/cocoa/AssertionCapability.h:
* Source/WebKit/Platform/cocoa/AssertionCapability.mm:
(WebKit::AssertionCapability::AssertionCapability):
(WebKit::AssertionCapability::platformCapability const):

* Source/WebKit/Platform/cocoa/MediaCapability.h: Added a ProcessCapability subclass for media
capabilities.
* Source/WebKit/Platform/cocoa/MediaCapability.mm:
(WebKit::createPlatformCapability):
(WebKit::MediaCapability::MediaCapability):
(WebKit::MediaCapability::environmentIdentifier const):

* Source/WebKit/Platform/cocoa/ProcessCapability.h:
(WebKit::ProcessCapability::environmentIdentifier const): Deleted.

* Source/WebKit/Platform/cocoa/ProcessCapabilityGrant.h: Added a class to represent a capability
grant created by ProcessCapabilityGranter.
(WebKit::ProcessCapabilityGrant::environmentIdentifier const):
* Source/WebKit/Platform/cocoa/ProcessCapabilityGrant.mm:
(WebKit::platformInvalidate):
(WebKit::ProcessCapabilityGrant::ProcessCapabilityGrant):
(WebKit::ProcessCapabilityGrant::~ProcessCapabilityGrant):
(WebKit::ProcessCapabilityGrant::isEmpty const):
(WebKit::ProcessCapabilityGrant::isValid const):
(WebKit::ProcessCapabilityGrant::setPlatformGrant):

* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h: Defined new _SECapabilities SPI.

* Source/WebKit/SourcesCocoa.txt:

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp: Stored a map of ProcessCapabilityGrants by
environmentIdentifier.
(WebKit::AuxiliaryProcessProxy::~AuxiliaryProcessProxy):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::processCapabilityGrants):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):

* Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.h: Added.
* Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm: Added.
(WebKit::granterQueue): Defined a queue for granting, revoking, and activating.
(WebKit::grantCapability): Internal helpers for granting on granterQueue.
(WebKit::grantCapabilityInternal): Ditto.
(WebKit::prepareGrant): Ditto.
(WebKit::finalizeGrant): Ditto.
(WebKit::ProcessCapabilityGranter::create):
(WebKit::ProcessCapabilityGranter::ProcessCapabilityGranter):
(WebKit::ProcessCapabilityGranter::grant): Grants a capability to the GPU and WebContent processes.
(WebKit::ProcessCapabilityGranter::revoke): Revokes a capability previously granted to the GPU and
WebContent processes.
(WebKit::ProcessCapabilityGranter::setMediaCapabilityActive): Activates a capability.
(WebKit::ProcessCapabilityGranter::invalidateGrants): Invalidates a vector of grants in a single
batch.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::mediaCapability const): Returns the current media capability or nullopt.
(WebKit::WebPageProxy::setMediaCapability): Sets or clears the current media capability.
(WebKit::WebPageProxy::updateMediaCapability): Updates the current media capability when the top
privately-controlled domain changes, and activates or deactivates it as needed.
(WebKit::WebPageProxy::shouldActivateMediaCapability const): Defines when a media capability should
be activated.
(WebKit::WebPageProxy::shouldDeactivateMediaCapability const): Defines when a media capability
should be deactivated.
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::processCapabilityGranter): Creates a ProcessCapabilityGranter if needed and
returns a reference to it.
(WebKit::WebProcessPool::gpuProcessForCapabilityGranter): Returns a GPUProcessProxy, if one exists,
to ProcessCapabilityGranter.
(WebKit::WebProcessPool::webProcessForCapabilityGranter): Returns a WebProcessProxy for a given
environmentIdentifier, if one exists, to ProcessCapabilityGranter.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm: Fixed a unified source issue.

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::gpuProcessExited): Invalidated existing grants when the GPU process exits.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchActivityStateChange): Called updateMediaCapability().
(WebKit::WebPageProxy::didCommitLoadForFrame): Ditto.
(WebKit::WebPageProxy::updatePlayingMediaDidChange): Ditto.
(WebKit::WebPageProxy::resetState): Set the mediaCapability to nullopt.
(WebKit::WebPageProxy::gpuProcessDidFinishLaunching): Granted an existing media capability to the
new GPU process.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processDidFinishLaunching): Granted existing media capabilities to the new
WebContent process.
(WebKit::WebProcessPool::disconnectProcess): Invalidated existing grants for the process that
disconnected.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/271761@main">https://commits.webkit.org/271761@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c629255ea2fd8c95c0a5d14d000da6fa1396329d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29573 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32092 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10388 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5528 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26787 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29846 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6875 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/25253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5876 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26341 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33436 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/25417 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26761 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32229 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/29789 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4156 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/30004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7700 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36144 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/7019 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Checked out pull request; Reviewed by Per Arne Vollan; compiling; layout-tests running") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6506 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7788 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3807 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm:
```diff
@@ -0,0 +1,305 @@
+            grantCapability(capability.get(), gpuExtension.get()),
+            grantCapability(capability.get(), webExtension.get())
+        };
+        return ProcessCapabilityGrantsPromise::createAndSettle(WTFMove(grants));
```
Wherever createAndSettle is used with invokeAsync you can directly return the value instead. So grants should do (but here will have to specify the lambda return type)

But I find `createAndResolve()` here clearer

jyavenard commented on Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm:
```diff
@@ -0,0 +1,305 @@
+
+static WorkQueue& granterQueue()
+{
+    static auto& queue = WorkQueue::create("ProcessCapabilityGranter Queue", WorkQueue::QOS::UserInitiated).leakRef();
```
Maybe use a Lazy init here. Also, that isn’t thread-safe with our compilation option (only used from the main thread so far, but it would be easy to forget that in future time)
```
    static std::once_flag once;
    static LazyNeverDestroyed<Ref<WorkQueue>> queue;
    std::call_once(once, [] {
        queue(WorkQueue::create("ProcessCapabilityGranter"));
    });
    return queue.get();
```

jyavenard commented on Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm:
```diff
@@ -0,0 +1,305 @@
+        needsWebProcessGrant
+    ](auto&& result) {
+        ProcessCapabilityGrant gpuProcessGrant { environmentIdentifier };
+        gpuProcessGrant.setPlatformGrant(result ? WTFMove(result.value().gpuProcessGrant) : nullptr);
```
`result->gnuProcessGrant`

no need for `value()` method, Expected has an `operator->` (and operator*())`

jyavenard commented on Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm:
```diff
@@ -0,0 +1,305 @@
+    bool needsGPUProcessGrant = gpuProcess ? prepareGrant(environmentIdentifier, *gpuProcess) : false;
+
+    RefPtr webProcess = m_client->webProcessForCapabilityGranter(*this, environmentIdentifier);
+    bool needsWebProcessGrant = webProcess ? prepareGrant(environmentIdentifier, *webProcess) : false;
```
`bool needsWebProcessGrant = webProcess && prepareGrant(environmentIdentifier, *webProcess)`

jyavenard commented on Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm:
```diff
@@ -0,0 +1,305 @@
+    }
+
+    RefPtr gpuProcess = m_client->gpuProcessForCapabilityGranter(*this);
+    bool needsGPUProcessGrant = gpuProcess ? prepareGrant(environmentIdentifier, *gpuProcess) : false;
```
`bool needsGPUProcessGrant = gpuProcess && prepareGrant(environmentIdentifier, *gpuProcess);`

jyavenard commented on Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm:
```diff
@@ -0,0 +1,305 @@
+    PlatformError,
+};
+
+using ProcessCapabilityGrantsPromise = NativePromise<ProcessCapabilityGrants, ProcessCapabilityGrantError>;
```
You are passing `ProcessCapabilityGrants` over different thread. It contains String. Are those string usable across threads? You'll probably need to call `isolatedCopy()` on them before passing them.

NativePromise can make the transfer safe, provided the `ProcessCapabilityGrants` has an `isolatedCopy()` method. If so, `ProcessCapabilityGrants ::isolatedCopy()` will be called before storing it internally and passing it to the final resolver.

aestes replied:
`ProcessCapabilityGrant` (singular) does contain a string, but `ProcessCapabilityGrants` is just a struct with two `RetainPtr`s. Maybe I should rename it to `PlatformProcessCapabilityGrants` to avoid confusion.

aestes replied:
However, I do pass a `Vector<ProcessCapabilityGrant>` across threads in `ProcessCapabilityGranter::invalidateGrants`. Oops, will fix.

pvollan commented on Source/WebKit/Platform/cocoa/AssertionCapability.mm:
```diff
@@ -32,27 +32,29 @@
 RetainPtr<_SECapabilities> AssertionCapability::platformCapability() const
 {
+#if USE(EXTENSIONKIT)
 #if USE(EXTENSIONKIT_INVALIDATION_CALLBACKS)
```
I think we may need to delay the enablement of the invalidation callbacks a few days.

aestes replied:
That's OK. This PR doesn't enable it (USE_EXTENSIONKIT_INVALIDATION_CALLBACKS is still false/0).

pvollan reviewed (approved):
R=me.

aestes committed:
[iOS] Grant media capabilities to the WebContent and GPU processes during capture and playback
https://bugs.webkit.org/show_bug.cgi?id=265588
rdar://116242683

Reviewed by Per Arne Vollan.

This patch introduces MediaCapability and ProcessCapabilityGranter, responsible for granting,
revoking, and managing the activation of a MediaCapability.

A MediaCapability is created each time a WebPageProxy navigates to a new top privately-controlled
domain, and is granted to the page's GPU and WebContent processes. When media playback or capture
becomes active on the page the capability is activated, and remains active for as long as that page
is capable of playback or capture. Granting and activating these capabilities allows iOS to know the
process hierarchy and domain responsible for playback and capture, and allows the system to, e.g.,
resume the responsible processes when starting playback from Now Playing or account for the
responsible domain in System Status. Media capabilities will serve as a replacement to SPIs like
AVSystemController_PIDToInheritApplicationStateFrom and -[AVAudioSession setHostProcessAttribution:].

This feature is behind an off-by-default web preference (MediaCapabilityGrantsEnabled). Adding API
tests is tracked by rdar://115750175. It wasn't done in this patch because tests are blocked by
other work.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Added a web preference.
* Source/WebKit/Platform/Logging.h: Added a logging channel.

* Source/WebKit/Platform/cocoa/AssertionCapability.h:
* Source/WebKit/Platform/cocoa/AssertionCapability.mm:
(WebKit::AssertionCapability::AssertionCapability):
(WebKit::AssertionCapability::platformCapability const):

* Source/WebKit/Platform/cocoa/MediaCapability.h: Added a ProcessCapability subclass for media
capabilities.
* Source/WebKit/Platform/cocoa/MediaCapability.mm:
(WebKit::createPlatformCapability):
(WebKit::MediaCapability::MediaCapability):
(WebKit::MediaCapability::environmentIdentifier const):

* Source/WebKit/Platform/cocoa/ProcessCapability.h:
(WebKit::ProcessCapability::environmentIdentifier const): Deleted.

* Source/WebKit/Platform/cocoa/ProcessCapabilityGrant.h: Added a class to represent a capability
grant created by ProcessCapabilityGranter.
(WebKit::ProcessCapabilityGrant::environmentIdentifier const):
* Source/WebKit/Platform/cocoa/ProcessCapabilityGrant.mm:
(WebKit::platformInvalidate):
(WebKit::ProcessCapabilityGrant::ProcessCapabilityGrant):
(WebKit::ProcessCapabilityGrant::~ProcessCapabilityGrant):
(WebKit::ProcessCapabilityGrant::isEmpty const):
(WebKit::ProcessCapabilityGrant::isValid const):
(WebKit::ProcessCapabilityGrant::setPlatformGrant):

* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h: Defined new _SECapabilities SPI.

* Source/WebKit/SourcesCocoa.txt:

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp: Stored a map of ProcessCapabilityGrants by
environmentIdentifier.
(WebKit::AuxiliaryProcessProxy::~AuxiliaryProcessProxy):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::processCapabilityGrants):
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):

* Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.h: Added.
* Source/WebKit/UIProcess/Cocoa/ProcessCapabilityGranter.mm: Added.
(WebKit::granterQueue): Defined a queue for granting, revoking, and activating.
(WebKit::grantCapability): Internal helpers for granting on granterQueue.
(WebKit::grantCapabilityInternal): Ditto.
(WebKit::prepareGrant): Ditto.
(WebKit::finalizeGrant): Ditto.
(WebKit::ProcessCapabilityGranter::create):
(WebKit::ProcessCapabilityGranter::ProcessCapabilityGranter):
(WebKit::ProcessCapabilityGranter::grant): Grants a capability to the GPU and WebContent processes.
(WebKit::ProcessCapabilityGranter::revoke): Revokes a capability previously granted to the GPU and
WebContent processes.
(WebKit::ProcessCapabilityGranter::setMediaCapabilityActive): Activates a capability.
(WebKit::ProcessCapabilityGranter::invalidateGrants): Invalidates a vector of grants in a single
batch.

* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::mediaCapability const): Returns the current media capability or nullopt.
(WebKit::WebPageProxy::setMediaCapability): Sets or clears the current media capability.
(WebKit::WebPageProxy::updateMediaCapability): Updates the current media capability when the top
privately-controlled domain changes, and activates or deactivates it as needed.
(WebKit::WebPageProxy::shouldActivateMediaCapability const): Defines when a media capability should
be activated.
(WebKit::WebPageProxy::shouldDeactivateMediaCapability const): Defines when a media capability
should be deactivated.
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::processCapabilityGranter): Creates a ProcessCapabilityGranter if needed and
returns a reference to it.
(WebKit::WebProcessPool::gpuProcessForCapabilityGranter): Returns a GPUProcessProxy, if one exists,
to ProcessCapabilityGranter.
(WebKit::WebProcessPool::webProcessForCapabilityGranter): Returns a WebProcessProxy for a given
environmentIdentifier, if one exists, to ProcessCapabilityGranter.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm: Fixed a unified source issue.

* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::gpuProcessExited): Invalidated existing grants when the GPU process exits.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::dispatchActivityStateChange): Called updateMediaCapability().
(WebKit::WebPageProxy::didCommitLoadForFrame): Ditto.
(WebKit::WebPageProxy::updatePlayingMediaDidChange): Ditto.
(WebKit::WebPageProxy::resetState): Set the mediaCapability to nullopt.
(WebKit::WebPageProxy::gpuProcessDidFinishLaunching): Granted an existing media capability to the
new GPU process.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::processDidFinishLaunching): Granted existing media capabilities to the new
WebContent process.
(WebKit::WebProcessPool::disconnectProcess): Invalidated existing grants for the process that
disconnected.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/271761@main
